Screenshot testing of examples on CI (#239)
pygfx/pygfx: README.md

Under development, many things can change.

## Testing examples

There are two types of tests for examples included with pygfx:

### Type 1: Checking if examples can run

> When running the test suite with `--slow` (enabled on CI), pytest will run every example in a subprocess,

Reviewer: The `--slow` is no more, right?

to see if it can run and exit cleanly. You can opt out of this mechanism by including the comment
`# run_example = false` in the module.

### Type 2: Checking if examples output an image

You can also (independently) opt-in to output testing for examples, by including the comment
`# test_example = true` in the module. Output testing means the test suite will

> attempt to import the `canvas` instance global from your example, and call `draw()` on it

Reviewer: Mmm, we're also hiding the canvas behind the renderer since #275. Can we use the global renderer instead?

Author: Well, I'm specifically relying on the OffscreenCanvas-only `draw()` method, which is not available on the Renderer. So, I'm not sure, what do you think?

Reviewer: ```py
im = example.renderer.target.draw()
```

Author: Done!

to see if an image is produced.

To support this type of testing, ensure the following requirements are met:

* The `WgpuCanvas` class is imported from the `wgpu.gui.auto` module.
* The `canvas` instance is exposed as a global in the module.
* A rendering callback has been registered with `canvas.request_draw(fn)`.

> Reference screenshot are available in the `examples/screenshots` folder,

Reviewer: ```suggestion
Reference screenshots are stored in the `examples/screenshots` folder,
```

the test suite will compare the rendered image with the reference.

> Note: this step will be skipped when not on running CI. Since images will have subtle differences
> depending on the system on which it was rendered, that would make the tests unreliable.

Reviewer: ```suggestion
Note: this step will be skipped when not running on CI. Since images will have subtle differences
depending on the system on which they are rendered, that would make the tests unreliable.
```
